src/Common/src/System/StringNormalizationExtensions.Unix.cs

```diff
@@ -6,35 +6,120 @@
 
 namespace System
 {
-    // TODO: We need an actual implementation here, but since all ASCII strings are normalized under all normalization forms
-    // we can handle the simple cases, which will actually allow a lot of code to just work.
-
     static partial class StringNormalizationExtensions
     {
         [SecurityCritical]
         public static bool IsNormalized(this string value, NormalizationForm normalizationForm)
         {
-            for (int i = 0; i < value.Length; i++)
+            if (value == null)
             {
-                if (value[i] > 0x7F)
-                {
-                    throw NotImplemented.ByDesign;
-                }
+                throw new ArgumentNullException("value");
+            }
+
+            if (normalizationForm != NormalizationForm.FormC && normalizationForm != NormalizationForm.FormD &&
+                normalizationForm != NormalizationForm.FormKC && normalizationForm != NormalizationForm.FormKD)
+            {
+                throw new ArgumentException(SR.Argument_InvalidNormalizationForm, "normalizationForm");
             }
 
-            return true;
+            if (HasInvalidUnicodeSequence(value))
+            {
+                throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, "value");
+            }
+
+            int ret = Interop.GlobalizationInterop.IsNormalized(normalizationForm, value, value.Length);
+
+            if (ret == -1)
+            {
+                throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, "value");
+            }
+
+            return ret == 1;
         }
 
         [SecurityCritical]
         public static string Normalize(this string value, NormalizationForm normalizationForm)
         {
-            if (IsNormalized(value, normalizationForm))
+            if (value == null)
+            {
+                throw new ArgumentNullException("value");
+            }
+
+            if (normalizationForm != NormalizationForm.FormC && normalizationForm != NormalizationForm.FormD &&
+                normalizationForm != NormalizationForm.FormKC && normalizationForm != NormalizationForm.FormKD)
+            {
+                throw new ArgumentException(SR.Argument_InvalidNormalizationForm, "normalizationForm");
+            }
+
+            if (HasInvalidUnicodeSequence(value))
+            {
+                throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, "value");
+            }
+
+            char[] buf = new char[value.Length];
```

Comment:
Correct.  Internally ICU will allocate a native buffer, normalize the string into it and then try to copy it to the buffer we pass in after normalization.  Doing things this way saves another round trip to native code and another native allocation, in the common case.